File: LeetCode/Study Plan/Practice Problems/Medium/Construct Binary Tree From Inorder And Postorder Traversal/Java/constructBinaryTreeFromInorderAndPostorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int[] inorder;
    int[] postorder;
    int postInd=0;
    HashMap<Integer, Integer> inorderIndex = new HashMap<>();
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        this.inorder = inorder;
        this.postorder = postorder;
        
        for (int i=0; i<inorder.length; i++) {
            inorderIndex.put(inorder[i], i);
        }

        postInd = postorder.length-1;
        return recur(0, inorder.length-1);
    }

    public TreeNode recur(int left, int right) {
        if (left > right) return null;
        
        TreeNode root = new TreeNode(this.postorder[this.postInd]);
        int ind = inorderIndex.get(root.val);
        this.postInd--;
        
        root.right = recur(ind+1, right);
        root.left = recur(left, ind-1);
        return root;
    }
}

File: LeetCode/Medium/Construct Binary Tree From Inorder And Postorder Traversal/Go/constructBinaryTreeFromInorderAndPostorder.go
package main

import "fmt"

type TreeNode struct {
	Val int
	Left,
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func New(val int) *TreeNode { return &TreeNode{Val: val} }

func buildTree(inorder []int, postorder []int) *TreeNode {
	inIndex := make(map[int]int)
	for i := 0; i < len(inorder); i++ {
		inIndex[inorder[i]] = i
	}
	postIndex := len(postorder) - 1
	return recur(0, len(inorder)-1, inorder, postorder, inIndex, &postIndex)
}

func recur(left, right int, inorder, postorder []int, inIndex map[int]int, postIndex *int) *TreeNode {
	if left > right {
		return nil
	}

	root := New(postorder[*postIndex])
	ind := inIndex[root.Val]
	*postIndex--
	root.Right = recur(ind+1, right, inorder, postorder, inIndex, postIndex)
	root.Left = recur(left, ind-1, inorder, postorder, inIndex, postIndex)
	return root
}

func main() {
	fmt.Println(buildTree([]int{9, 3, 15, 20, 7}, []int{9, 15, 7, 20, 3}))
}
